File: vsg_qt/add_job_dialog/ui.py
# vsg_qt/add_job_dialog/ui.py
from __future__ import annotations

from PySide6.QtWidgets import (
    QDialog,
    QDialogButtonBox,
    QFileDialog,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QMessageBox,
    QPushButton,
    QScrollArea,
    QVBoxLayout,
    QWidget,
)

from vsg_core.job_discovery import discover_jobs


class SourceInputWidget(QWidget):
    """A self-contained widget for a single source input row that handles drag-and-drop."""
    def __init__(self, source_num: int, parent=None):
        super().__init__(parent)
        self.setAcceptDrops(True)

        layout = QHBoxLayout(self)

        label_text = f"Source {source_num} (Reference):" if source_num == 1 else f"Source {source_num}:"
        label = QLabel(label_text)
        self.line_edit = QLineEdit()
        browse_btn = QPushButton("Browse…")

        browse_btn.clicked.connect(self._browse_for_path)

        layout.addWidget(label, 1)
        layout.addWidget(self.line_edit, 4)
        layout.addWidget(browse_btn)

    def dragEnterEvent(self, event):
        if event.mimeData().hasUrls():
            event.acceptProposedAction()
        else:
            event.ignore()

    def dropEvent(self, event):
        if event.mimeData().hasUrls():
            # Use the path of the first dropped file
            path = event.mimeData().urls()[0].toLocalFile()
            self.line_edit.setText(path)
            event.acceptProposedAction()
        else:
            event.ignore()

    def _browse_for_path(self):
        dialog = QFileDialog(self, "Select Source")
        dialog.setFileMode(QFileDialog.FileMode.AnyFile)
        if dialog.exec():
            self.line_edit.setText(dialog.selectedFiles()[0])

    def text(self) -> str:
        return self.line_edit.text()

class AddJobDialog(QDialog):
    """
    A dialog for dynamically adding sources to discover jobs.
    """
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setWindowTitle("Add Job(s) to Queue")
        self.setMinimumSize(700, 300)

        self.discovered_jobs: list[dict] = []
        self.source_widgets: list[SourceInputWidget] = []
        self._build_ui()

        # Start with 2 sources by default
        self.add_source_input()
        self.add_source_input()

    def _build_ui(self):
        layout = QVBoxLayout(self)

        scroll_area = QScrollArea()
        scroll_area.setWidgetResizable(True)

        container = QWidget()
        self.inputs_layout = QVBoxLayout(container)
        scroll_area.setWidget(container)

        layout.addWidget(scroll_area)

        add_source_btn = QPushButton("Add Another Source")
        add_source_btn.clicked.connect(self.add_source_input)
        layout.addWidget(add_source_btn)

        dialog_btns = QDialogButtonBox(QDialogButtonBox.Ok | QDialogButtonBox.Cancel)
        ok_button = dialog_btns.button(QDialogButtonBox.Ok)
        ok_button.setText("Find & Add Jobs")

        ok_button.clicked.connect(self.find_and_accept)
        dialog_btns.rejected.connect(self.reject)

        layout.addWidget(dialog_btns)

    def add_source_input(self):
        """Adds a new SourceInputWidget to the dialog."""
        source_num = len(self.source_widgets) + 1
        source_widget = SourceInputWidget(source_num)
        self.source_widgets.append(source_widget)
        self.inputs_layout.addWidget(source_widget)

    def populate_sources_from_paths(self, paths: list[str]):
        """Pre-fills the source inputs from a list of paths."""
        # Clear any default inputs
        while self.inputs_layout.count():
            child = self.inputs_layout.takeAt(0)
            if child and child.widget():
                child.widget().deleteLater()
        self.source_widgets.clear()

        # Add an input for each dropped path
        for path in paths:
            self.add_source_input()
            self.source_widgets[-1].line_edit.setText(path)

        # Ensure at least two inputs exist if user drops only one file
        if len(self.source_widgets) < 2:
            self.add_source_input()

    def find_and_accept(self):
        """Discover jobs from paths and accept the dialog if any are found."""
        sources: dict[str, str] = {}
        for i, source_widget in enumerate(self.source_widgets):
            path = source_widget.text().strip()
            if path:
                sources[f"Source {i+1}"] = path

        if "Source 1" not in sources:
            QMessageBox.warning(self, "Input Required", "Source 1 (Reference) cannot be empty.")
            return

        try:
            self.discovered_jobs = discover_jobs(sources)
            if not self.discovered_jobs:
                QMessageBox.information(self, "No Jobs Found", "No matching jobs could be discovered from the provided paths.")
                return

            self.accept()
        except (ValueError, FileNotFoundError) as e:
            QMessageBox.critical(self, "Error Discovering Jobs", str(e))

    def get_discovered_jobs(self) -> list[dict]:
        return self.discovered_jobs

File: qt_ui/add_job_dialog/ui.cpp
// Add Job Dialog Implementation

#include "ui.hpp"

#include <QHBoxLayout>
#include <QVBoxLayout>
#include <QLabel>
#include <QPushButton>
#include <QDialogButtonBox>
#include <QScrollArea>
#include <QFileDialog>
#include <QMessageBox>
#include <QDragEnterEvent>
#include <QDropEvent>
#include <QMimeData>
#include <QUrl>

// =============================================================================
// SourceInputWidget
// =============================================================================

SourceInputWidget::SourceInputWidget(int sourceNum, QWidget* parent)
    : QWidget(parent)
{
    setAcceptDrops(true);

    auto* layout = new QHBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);

    QString labelText = (sourceNum == 1)
        ? QString("Source %1 (Reference):").arg(sourceNum)
        : QString("Source %1:").arg(sourceNum);

    auto* label = new QLabel(labelText);
    label->setMinimumWidth(120);

    m_lineEdit = new QLineEdit();

    auto* browseBtn = new QPushButton("Browse...");
    connect(browseBtn, &QPushButton::clicked, this, &SourceInputWidget::browseForPath);

    layout->addWidget(label, 1);
    layout->addWidget(m_lineEdit, 4);
    layout->addWidget(browseBtn);
}

void SourceInputWidget::dragEnterEvent(QDragEnterEvent* event)
{
    if (event->mimeData()->hasUrls()) {
        event->acceptProposedAction();
    } else {
        event->ignore();
    }
}

void SourceInputWidget::dropEvent(QDropEvent* event)
{
    if (event->mimeData()->hasUrls()) {
        QList<QUrl> urls = event->mimeData()->urls();
        if (!urls.isEmpty()) {
            m_lineEdit->setText(urls.first().toLocalFile());
        }
        event->acceptProposedAction();
    } else {
        event->ignore();
    }
}

void SourceInputWidget::browseForPath()
{
    QFileDialog dialog(this, "Select Source");
    dialog.setFileMode(QFileDialog::AnyFile);
    dialog.setNameFilter("Video Files (*.mkv *.mp4 *.avi *.m4v *.mov *.ts);;All Files (*)");

    if (!m_lineEdit->text().isEmpty()) {
        QFileInfo fi(m_lineEdit->text());
        dialog.setDirectory(fi.absolutePath());
    }

    if (dialog.exec() == QDialog::Accepted) {
        QStringList files = dialog.selectedFiles();
        if (!files.isEmpty()) {
            m_lineEdit->setText(files.first());
        }
    }
}

// =============================================================================
// AddJobDialog
// =============================================================================

AddJobDialog::AddJobDialog(QWidget* parent)
    : QDialog(parent)
{
    setWindowTitle("Add Job(s) to Queue");
    setMinimumSize(700, 300);

    buildUi();

    // Start with 2 sources by default
    addSourceInput();
    addSourceInput();
}

void AddJobDialog::buildUi()
{
    auto* mainLayout = new QVBoxLayout(this);

    // Scroll area for source inputs
    auto* scrollArea = new QScrollArea();
    scrollArea->setWidgetResizable(true);

    auto* container = new QWidget();
    m_inputsLayout = new QVBoxLayout(container);
    scrollArea->setWidget(container);

    mainLayout->addWidget(scrollArea);

    // Add source button
    auto* addSourceBtn = new QPushButton("Add Another Source");
    connect(addSourceBtn, &QPushButton::clicked, this, &AddJobDialog::addSourceInput);
    mainLayout->addWidget(addSourceBtn);

    // Dialog buttons
    auto* buttonBox = new QDialogButtonBox(
        QDialogButtonBox::Ok | QDialogButtonBox::Cancel);

    auto* okBtn = buttonBox->button(QDialogButtonBox::Ok);
    okBtn->setText("Find && Add Jobs");

    connect(okBtn, &QPushButton::clicked, this, &AddJobDialog::findAndAccept);
    connect(buttonBox, &QDialogButtonBox::rejected, this, &QDialog::reject);

    mainLayout->addWidget(buttonBox);
}

void AddJobDialog::addSourceInput()
{
    int sourceNum = static_cast<int>(m_sourceWidgets.size()) + 1;
    auto* widget = new SourceInputWidget(sourceNum);
    m_sourceWidgets.push_back(widget);
    m_inputsLayout->addWidget(widget);
}

void AddJobDialog::populateSourcesFromPaths(const QStringList& paths)
{
    // Clear existing inputs
    while (m_inputsLayout->count() > 0) {
        QLayoutItem* item = m_inputsLayout->takeAt(0);
        if (item && item->widget()) {
            item->widget()->deleteLater();
        }
        delete item;
    }
    m_sourceWidgets.clear();

    // Add an input for each path
    for (const QString& path : paths) {
        addSourceInput();
        m_sourceWidgets.back()->setText(path);
    }

    // Ensure at least two inputs
    while (m_sourceWidgets.size() < 2) {
        addSourceInput();
    }
}

void AddJobDialog::findAndAccept()
{
    // Collect non-empty source paths
    std::map<QString, QString> sources;
    for (size_t i = 0; i < m_sourceWidgets.size(); ++i) {
        QString path = m_sourceWidgets[i]->text().trimmed();
        if (!path.isEmpty()) {
            sources[QString("Source %1").arg(i + 1)] = path;
        }
    }

    // Validate
    if (sources.find("Source 1") == sources.end()) {
        QMessageBox::warning(this, "Input Required",
            "Source 1 (Reference) cannot be empty.");
        return;
    }

    if (sources.size() < 2) {
        QMessageBox::warning(this, "Input Required",
            "At least two sources are required to discover jobs.");
        return;
    }

    // TODO: Call bridge to discover jobs
    // auto jobs = vsg::bridge_discover_jobs(paths);

    // For now, create a stub job from the inputs
    m_discoveredJobs.clear();
    m_discoveredJobs.push_back(sources);

    accept();
}

std::vector<std::map<QString, QString>> AddJobDialog::getDiscoveredJobs() const
{
    return m_discoveredJobs;
}
